src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,206 @@
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addNodeToStart(T value) {
+        Node<T> newNode = new Node<>(first, value, null);
+        first.prev = newNode;
+        first = newNode;
+        size++;
+    }
+
+    private void addNoteToEnd(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        last = newNode;
+        size++;
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private void checkElementIndex(int index) {
+
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException(
+                    "Index out of bounds exception! Please enter a correct value!");
+        }
+    }
+
+    private void checkElementIndexForAdd(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(
+                    "Index out of bounds exception! Please enter a correct value!");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode;
+        int steps;
+
+        if (index <= midIndex) {
+            currentNode = first;
+            steps = index;
+        } else {
+            currentNode = last;
+            steps = size - 1 - index;
+        }
+        for (int i = 0; i < steps; i++) {
+            currentNode = (index <= midIndex) ? currentNode.next : currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private T unlink(Node<T> x) {
+        final T element = x.element;
+        final Node<T> next = x.next;
+        final Node<T> prev = x.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+
+        x.element = null;
+        size--;
+        return element;
+    }
+
+    private Node<T> node(int index) {
+
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    public void addFirstNode(T element) {
+        first = new Node<>(null, element, null);
+        last = first;
+        size++;
+    }
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
```

Comment:
Don't use getters and setters in class Node. Since Node is an inner class, we have access to its private fields directly.